lib/fast_jsonapi/relationship.rb

```diff
@@ -136,13 +158,39 @@ def id_hash(id, record_type, default_return = false)
     end
 
     def fetch_id(record, params)
-      if object_block.present?
-        object = FastJsonapi.call_proc(object_block, record, params)
-        return object.map { |item| item.public_send(id_method_name) } if object.respond_to? :map
+      if object_block.nil?
+        #
+        # TODO: This line assumes the `id_method_name` is a foreign key defined on the parent record.
+        #       The `id_method_name` option should be replaced with `foreign_key` to make the intent
+        #       explicit.
+        #
+        foreign_key = id_method_name
+        foreign_key ||= (has_many? ? "#{@name.to_s.singularize}_ids" : "#{@name}_id")
```

Comment:
For performance, we should memoize this computation, as we may be iterating over hundreds of records or more, and it is constant once the relationship is defined.

```ruby
@computed_id_method_name = has_many? ? "#{@name.to_s.singularize}_ids" : "#{@name}_id"
```